### PR TITLE
Fixes #22150: API for directive compliance

### PR DESCRIPTION
**Pull request by ElaadF**
https://issues.rudder.io/issues/22150

### CODE REVIEW

**Comment by ElaadF** [comment]
Commit modified

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/data/Compliance.scala [hunk 1]
```diff
@@ -301,6 +321,300 @@ object JsonCompliance {
...
+     */
+
+    def toJson(level: Int, precision: CompliancePrecision) = (
+      ("id"                    -> directive.id.serialize)
+        ~ ("name"              -> directive.name)
+        ~ ("compliance"        -> directive.compliance.complianceWithoutPending(precision))
+        ~ ("mode"              -> directive.mode.name)
+        ~ ("complianceDetails" -> percents(directive.compliance, precision))
+//        ~ ("rules"             -> rules(directive.rules.map(_.), level, precision))
+        ~ ("nodes"             -> byNodes(directive.nodes, level, precision))
```
Comment (ElaadF) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/data/Compliance.scala:348]: I'm trying to get a structure like
```
"nodes": [
          {
            "id": "831b9c93-bb96-4d48-8c4e-91ca19b98bde",
            "name": "prod-zabbix-01.lab.rudder.io",
            "compliance": 100.0,
            "complianceDetails": {
              "auditCompliant": 100.0
            },
            "rules": [
              {
                "id": "94e9b2ed-6f5a-44a2-a8c5-82b88f4d85b7",
                "name": "Training - Trigger Command",
                "compliance": 100.0,
                "complianceDetails": {
                  "auditCompliant": 100.0
                },
                "components": [
                  {
                    "name": "Copy script agent-hello onto Nodes ",
                    "compliance": 100.0,
                    "complianceDetails": {
                      "auditCompliant": 100.0
                    },
                    "values": [
                      {
                        "value": "/opt/rudder/share/commands/agent-hello",
                        "reports": [
                          {
                            "status": "auditCompliant",
                            "message": "Copying /opt/rudder/share/commands/agent-hello from agent-hello was correct"
                          }
                        ]
                      }
                    ]
                  }
                ]
```

Comment (ncharles) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/data/Compliance.scala:348]: why do you have a rule and not a directive?

Comment (VinceMacBuche) [reply webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/data/Compliance.scala:348]: directive is at top level. we want compliance for a directive

**Comment by ElaadF** [comment]
Commit modified

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala [hunk 2]
```diff
@@ -469,6 +519,34 @@ class ComplianceAPIService(
...
+
+      compliance   <- getGlobalComplianceMode().toIO
+
+    } yield {
+      ByDirectiveRuleCompliance(
+        directiveId,
+        directive.map(_.id.uid.value).getOrElse("Unknown"),
+        ComplianceLevel.sum(byRules.map(_.compliance)),
+        compliance.mode,
+        byRules,
```
Comment (ElaadF) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/lift/ComplianceApi.scala:544]: Here we expect `Seq[ByDirectiveByRuleComponentCompliance]` but `getByRulesCompliance` return `Seq[ByRuleRuleCompliance]`

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
Commit modified

**Comment by ElaadF** [comment]
Commit modified

**Review by VinceMacBuche — APPROVED** [review]

File: webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala [hunk 3]
```diff
@@ -160,6 +160,13 @@ object ComplianceApi       extends ApiModuleProvider[ComplianceApi] {
     val dataContainer  = Some("globalCompliance")
   }
 
+  final case object GetDirectiveComplianceId extends ComplianceApi with OneParam with StartsAtVersion10 with SortIndex {
```
Comment (ncharles) [inline webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/EndpointsDefinition.scala:163]: why does it start at version 10 and not latest version ?

**Pull request merged**